### PR TITLE
fix: various alembic migration issues with queries

### PR DESCRIPTION
**Pull request by michael-genson**
## What type of PR is this?

_(REQUIRED)_

- bug

## What this PR does / why we need it:

_(REQUIRED)_

Building off of https://github.com/mealie-recipes/mealie/pull/2760, another issue came up where committing the transaction was executing the faulty SELECT statement. This PR should prevent this by disabling auto refreshes after commits for this session.

## Which issue(s) this PR fixes:

_(REQUIRED)_

Mentioned in https://github.com/mealie-recipes/mealie/issues/2769

## Testing

_(fill-in or delete this section)_

Pulled a backup from an older version with dupes and confirmed it migrates correctly now.

### CODE REVIEW

**Comment by zierbeek** [comment]
could you maybe give my backup a test?

**Comment by michael-genson** [comment]
Sure, do you have a link?

**Comment by zierbeek** [comment]
[Uploading mealie_2023.11.10.14.35.55.zip…]()

Yes!

**Comment by michael-genson** [comment]
Might need to re-upload, GitHub doesn't seem to like it. It also might be too big for GH:

image

**Comment by zierbeek** [comment]
[Uploading mealie_2023.11.10.14.35.55.zip…]()

**Comment by michael-genson** [comment]
Whelp, your backup broke several things. They seem fixed now, though to be honest I'm not quite sure why. I built the SQL manually and ran some tests locally to make sure data was still intact. I'm going to send over the migrated db on Discord, if you can check to make sure it works that'd be great.

File: alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py [hunk 1]
```diff
@@ -33,21 +33,29 @@ def populate_normalized_fields():
     )
     for unit in units:
         if unit.name is not None:
-            unit.name_normalized = IngredientUnitModel.normalize(unit.name)
+            session.execute(
+                sa.text(
+                    f"UPDATE {IngredientUnitModel.__tablename__} SET name_normalized='{IngredientUnitModel.normalize(unit.name)}' WHERE id='{unit.id}'"
```
Comment (hay-kot) [inline alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py]: FYI, you either need to use a parameterized query or properly escape the text to ensure it doesn't break stuff. 

See #2783 for an example of the problem.

File: alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py [hunk 2]
```diff
@@ -33,21 +33,29 @@ def populate_normalized_fields():
...
+                )
+            )
 
         if unit.abbreviation is not None:
-            unit.abbreviation_normalized = IngredientUnitModel.normalize(unit.abbreviation)
-
-        session.add(unit)
+            session.execute(
+                sa.text(
+                    f"UPDATE {IngredientUnitModel.__tablename__} SET abbreviation_normalized='{IngredientUnitModel.normalize(unit.abbreviation)}' WHERE id='{unit.id}'"
```
Comment (hay-kot) [inline alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py]: See above

File: alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py [hunk 3]
```diff
@@ -33,21 +33,29 @@ def populate_normalized_fields():
...
         session.execute(select(IngredientFoodModel).options(orm.load_only(IngredientFoodModel.name))).scalars().all()
     )
     for food in foods:
         if food.name is not None:
-            food.name_normalized = IngredientFoodModel.normalize(food.name)
-
-        session.add(food)
+            session.execute(
+                sa.text(
+                    f"UPDATE {IngredientFoodModel.__tablename__} SET name_normalized='{IngredientFoodModel.normalize(food.name)}' WHERE id='{food.id}'"
```
Comment (hay-kot) [inline alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py]: See above

File: alembic/versions/2023-10-04-14.29.26_dded3119c1fe_added_unique_constraints.py [hunk 4]
```diff
@@ -71,62 +65,39 @@ def _resolve_duplicate_food(
         .all()
     ):
         recipe_ingredient.food_id = keep_food_id
-        recipe_ingredient.food = keep_food
 
-    session.commit()
-
-    session.query(IngredientFoodModel).options(load_only(IngredientFoodModel.id)).filter_by(id=dupe_food_id).delete()
-    session.commit()
+    session.execute(sa.text(f"DELETE FROM {IngredientFoodModel.__tablename__} WHERE id='{dupe_food_id}'"))
```
Comment (hay-kot) [inline alembic/versions/2023-10-04-14.29.26_dded3119c1fe_added_unique_constraints.py]: See above

File: alembic/versions/2023-10-04-14.29.26_dded3119c1fe_added_unique_constraints.py [hunk 5]
```diff
@@ -71,62 +65,39 @@ def _resolve_duplicate_food(
...
 
     for recipe_ingredient in session.query(RecipeIngredientModel).filter_by(unit_id=dupe_unit_id).all():
         recipe_ingredient.unit_id = keep_unit_id
-        recipe_ingredient.unit = keep_unit
 
-    session.commit()
-
-    session.query(IngredientUnitModel).options(load_only(IngredientUnitModel.id)).filter_by(id=dupe_unit_id).delete()
-    session.commit()
+    session.execute(sa.text(f"DELETE FROM {IngredientUnitModel.__tablename__} WHERE id='{dupe_unit_id}'"))
```
Comment (hay-kot) [inline alembic/versions/2023-10-04-14.29.26_dded3119c1fe_added_unique_constraints.py]: See above

File: alembic/versions/2023-10-04-14.29.26_dded3119c1fe_added_unique_constraints.py [hunk 6]
```diff
@@ -71,62 +65,39 @@ def _resolve_duplicate_food(
...
-        shopping_list_item.label = keep_label
-
-    session.commit()
 
     for ingredient_food in session.query(IngredientFoodModel).filter_by(label_id=dupe_label_id).all():
         ingredient_food.label_id = keep_label_id
-        ingredient_food.label = keep_label
 
-    session.commit()
+    session.execute(sa.text(f"DELETE FROM {MultiPurposeLabel.__tablename__} WHERE id='{dupe_label_id}'"))
```
Comment (hay-kot) [inline alembic/versions/2023-10-04-14.29.26_dded3119c1fe_added_unique_constraints.py]: See above

**Review by hay-kot — CHANGES_REQUESTED** [review]

File: alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py (hunk 1 above)
Comment (michael-genson) [reply alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py]: I thought the normalize function we're running is supposed to strip all special characters from the string, but I guess it's not taking apostrophes out. That might be a problem for searches.

Either way we should be doing it regardless, so I'll go ahead and add that, thanks!

File: alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py [hunk 7]
```diff
@@ -33,21 +33,29 @@ def populate_normalized_fields():
     )
     for unit in units:
         if unit.name is not None:
-            unit.name_normalized = IngredientUnitModel.normalize(unit.name)
+            session.execute(
+                sa.text(
+                    f"UPDATE {IngredientUnitModel.__tablename__} SET name_normalized=:name_normalized WHERE id=:id"
```
Comment (hay-kot) [inline alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py:38]: Do these substitutions work in both Postgres and SQL? I can't remember if they are db agnostic or not.

Comment (michael-genson) [reply alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py:38]: It looks like it works in Postgres: https://www.postgresql.org/docs/current/app-psql.html#APP-PSQL-INTERPOLATION

This stackoverflow answer says it should work with psycopg2 (which we use, I'm pretty sure): https://stackoverflow.com/a/39414254

And it definitely works in SQLite (I tested it locally but also: https://www.sqlite.org/c3ref/bind_blob.html)

Comment (michael-genson) [reply alembic/versions/2023-09-01-14.55.42_0341b154f79a_added_normalized_unit_and_food_names.py:38]: Although if someone can verify on postgres that'd be great

**Comment by michael-genson** [comment]
I had two different sqlite formats in my test data - one had UUIDs with dashes, one didn't. I resolved the ambiguity by changing the select statement to just give me the raw DB value, rather than trying to coerce things. Both of my test dbs were successful with this change

**Review by hay-kot — APPROVED** [review]
LGTM, will leave it up to you to merge 🚀

**Pull request merged**